File: web/src/lib/tip.ts
export type TipChannelKey = "wechat" | "alipay" | "custom";

export interface TipChannel {
  key: TipChannelKey;
  label: string;
  qrImage: string;
  link: string;
}

export interface TipConfig {
  enabled: boolean;
  title: string;
  message: string;
  channels: TipChannel[];
}

export const MAX_TIP_CHANNELS = 4;
export const MAX_TIP_TITLE_LEN = 32;
export const MAX_TIP_MESSAGE_LEN = 120;
export const MAX_TIP_LABEL_LEN = 32;
export const MAX_TIP_IMAGE_LEN = 512;
export const MAX_TIP_LINK_LEN = 512;
export const DEFAULT_TIP_TITLE = "赞赏支持";
export const DEFAULT_TIP_MESSAGE = "如果这个站对你有帮助，欢迎请作者喝杯咖啡";

export function createDefaultTipConfig(): TipConfig {
  return {
    enabled: false,
    title: DEFAULT_TIP_TITLE,
    message: DEFAULT_TIP_MESSAGE,
    channels: [
      { key: "wechat", label: "微信", qrImage: "", link: "" },
      { key: "alipay", label: "支付宝", qrImage: "", link: "" },
    ],
  };
}

function isTipChannelKey(value: string): value is TipChannelKey {
  return value === "wechat" || value === "alipay" || value === "custom";
}

function defaultTipLabel(key: TipChannelKey): string {
  if (key === "wechat") return "微信";
  if (key === "alipay") return "支付宝";
  return "自定义";
}

function truncateChars(raw: string, max: number): string {
  const chars = Array.from(raw);
  if (max <= 0 || chars.length <= max) return raw;
  return chars.slice(0, max).join("");
}

function normalizeTipLink(raw: string): string {
  const link = raw.trim();
  if (!link) return "";
  const lower = link.toLowerCase();
  if (lower.startsWith("http://") || lower.startsWith("https://")) {
    return truncateChars(link, MAX_TIP_LINK_LEN);
  }
  if (link.includes(":") || link.startsWith("//")) {
    return "";
  }
  if (link.includes(".") && !link.includes(" ")) {
    return truncateChars(`https://${link.replace(/^\/+/, "")}`, MAX_TIP_LINK_LEN);
  }
  return "";
}

function normalizeTipImage(raw: string): string {
  const src = raw.trim();
  if (!src) return "";
  const lower = src.toLowerCase();
  if (lower.startsWith("http://") || lower.startsWith("https://") || src.startsWith("/")) {
    return truncateChars(src, MAX_TIP_IMAGE_LEN);
  }
  return "";
}

export function normalizeTipConfig(input?: Partial<TipConfig> | null): TipConfig {
  const title = truncateChars(String(input?.title ?? "").trim() || DEFAULT_TIP_TITLE, MAX_TIP_TITLE_LEN);
  const message = truncateChars(
    String(input?.message ?? "").trim() || DEFAULT_TIP_MESSAGE,
    MAX_TIP_MESSAGE_LEN,
  );
  const channels: TipChannel[] = [];

  for (const raw of input?.channels ?? []) {
    const keyRaw = String(raw?.key ?? "").trim();
    const key: TipChannelKey = isTipChannelKey(keyRaw) ? keyRaw : "custom";
    const rawLabel = String(raw?.label ?? "").trim();
    const qrImage = normalizeTipImage(String(raw?.qrImage ?? ""));
    const link = normalizeTipLink(String(raw?.link ?? ""));
    if (!qrImage && !link && key === "custom" && !rawLabel) {
      continue;
    }
    channels.push({
      key,
      label: truncateChars(rawLabel || defaultTipLabel(key), MAX_TIP_LABEL_LEN),
      qrImage,
      link,
    });
    if (channels.length >= MAX_TIP_CHANNELS) {
      break;
    }
  }

  return {
    enabled: Boolean(input?.enabled),
    title,
    message,
    channels: channels.length > 0 ? channels : createDefaultTipConfig().channels,
  };
}

export function serializeTipConfig(tip?: Partial<TipConfig> | null): string {
  return JSON.stringify(normalizeTipConfig(tip));
}

export function visibleTipChannels(tip?: TipConfig | null): TipChannel[] {
  return (tip?.channels ?? []).filter(
    (channel) => Boolean(channel.qrImage?.trim()) || Boolean(channel.link?.trim()),
  );
}

export function hasVisibleTip(tip?: TipConfig | null): boolean {
  return Boolean(tip?.enabled) && visibleTipChannels(tip).length > 0;
}

File: server/internal/model/tip.go
package model

import (
	"encoding/json"
	"strings"
	"unicode/utf8"
)

func defaultTipLabel(key string) string {
	switch key {
	case TipChannelWeChat:
		return "微信"
	case TipChannelAlipay:
		return "支付宝"
	default:
		return "自定义"
	}
}

func truncateRunes(raw string, max int) string {
	if max <= 0 || utf8.RuneCountInString(raw) <= max {
		return raw
	}
	return string([]rune(raw)[:max])
}

func normalizeTipLink(raw string) string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return ""
	}
	lower := strings.ToLower(raw)
	if strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://") {
		return truncateRunes(raw, MaxTipLinkLen)
	}
	if strings.Contains(raw, ":") || strings.HasPrefix(raw, "//") {
		return ""
	}
	if strings.Contains(raw, ".") && !strings.Contains(raw, " ") {
		return truncateRunes("https://"+strings.TrimLeft(raw, "/"), MaxTipLinkLen)
	}
	return ""
}

func normalizeTipImage(raw string) string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return ""
	}
	lower := strings.ToLower(raw)
	if strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://") || strings.HasPrefix(raw, "/") {
		return truncateRunes(raw, MaxTipImageLen)
	}
	return ""
}

func isTipChannelKey(key string) bool {
	return key == TipChannelWeChat || key == TipChannelAlipay || key == TipChannelCustom
}

func normalizeTipChannelKey(raw string) string {
	key := strings.TrimSpace(raw)
	if isTipChannelKey(key) {
		return key
	}
	return TipChannelCustom
}

// NormalizeTipConfig 清洗赞赏配置：补默认文案、限制渠道数、丢弃无效项。
func NormalizeTipConfig(tip TipConfig) TipConfig {
	title := strings.TrimSpace(tip.Title)
	if title == "" {
		title = DefaultTipTitle
	}
	title = truncateRunes(title, MaxTipTitleLen)
	message := strings.TrimSpace(tip.Message)
	if message == "" {
		message = DefaultTipMessage
	}
	message = truncateRunes(message, MaxTipMessageLen)

	channels := make([]TipChannel, 0, len(tip.Channels))
	for _, ch := range tip.Channels {
		key := normalizeTipChannelKey(ch.Key)
		rawLabel := strings.TrimSpace(ch.Label)
		label := rawLabel
		if label == "" {
			label = defaultTipLabel(key)
		}
		label = truncateRunes(label, MaxTipLabelLen)
		qr := normalizeTipImage(ch.QrImage)
		link := normalizeTipLink(ch.Link)
		if qr == "" && link == "" && key == TipChannelCustom && rawLabel == "" {
			continue
		}
		channels = append(channels, TipChannel{
			Key:     key,
			Label:   label,
			QrImage: qr,
			Link:    link,
		})
		if len(channels) >= MaxTipChannels {
			break
		}
	}
	if len(channels) == 0 {
		channels = DefaultTipConfig().Channels
	}

	return TipConfig{
		Enabled:  tip.Enabled,
		Title:    title,
		Message:  message,
		Channels: channels,
	}
}

// EncodeTipJSON 将赞赏配置编码为持久化 JSON。
func EncodeTipJSON(tip TipConfig) string {
	raw, err := json.Marshal(NormalizeTipConfig(tip))
	if err != nil {
		return ""
	}
	return string(raw)
}

// DecodeTipJSON 解析赞赏配置 JSON，非法或空值回退默认。
func DecodeTipJSON(raw string) TipConfig {
	if strings.TrimSpace(raw) == "" {
		return DefaultTipConfig()
	}
	var tip TipConfig
	if err := json.Unmarshal([]byte(raw), &tip); err != nil {
		return DefaultTipConfig()
	}
	return NormalizeTipConfig(tip)
}
